scw: don't fetch dependencies in build phase

amake opened the pull request:
#### Description

Per https://trac.macports.org/ticket/61192 this is one of the golang ports that automatically downloads its dependencies at build time.

To fix it, I used `go2port`. There were no unexpected difficulties.

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 10.15.6 19G2021
Xcode 12.0 12A7209

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

dgsb reviewed (commented):
As a global note, I'm not sure to understand the advantage of this method.
In the end we still have to download the dependencies to build the package.
It looks like we are reimplementing a dependency management at macports level.
It seems it's going to make package version bump harder to handle.

amake commented:
> As a global note, I'm not sure to understand the advantage of this method.

I addressed that here: https://trac.macports.org/ticket/61192#comment:55

> In the end we still have to download the dependencies to build the package.

Yes but this approach allows MacPorts to mirror distfiles and ensure they don't change from one build to the next. 

> It looks like we are reimplementing a dependency management at macports level.

This is well established MacPorts practice. See for example Rust ports using the cargo portgroup(s).

Of course the current `go.vendors` mechanism is not perfect. Improvements or alternatives that meet the requirements are welcome. 

> It seems it's going to make package version bump harder to handle.

The `go2port` tool has an (imperfect) function for updating an existing portfile.

amake commented on net/scw/Portfile:
```diff
@@ -18,11 +18,170 @@ description         Command Line Interface for Scaleway
+                        size    1566580
 
-build.args          ./cmd/scw
+go.vendors          github.com/pkg/errors \
```
For technical reasons, the preferred `go.vendors` order is _reverse_ alphabetical (though I only implemented that in `go2port` _after_ creating this PR).

I can change it to reverse-alphabetical order if you really want it, but I won't do alphabetical.

dgsb commented:
Won't all these issues be solved by setting appropriatly the GOPROXY env var to the canonical https://proxy.golang.org/ ?
It would allow to keep a clean and modern module based build

dgsb commented:
Or macports team could add a their own goproxy instance in their infrastructure ?

dgsb commented:
If I understand this doc correctly, we should already use this proxy. Or am I wrong ?
https://golang.org/cmd/go/#hdr-Module_downloading_and_verification

amake commented:
Honestly I don't know, but I have seen for instance in https://trac.macports.org/ticket/61185 a dependency served from a defunct domain caused building to fail. So if we are using the default proxy then it's not solving one of the major problems that the MacPorts distfile fetching policy is designed to solve. 

Running our own goproxy could be another way to solve it, but right now we don't have one.

dgsb commented:
@amake it seems that kubergrunt was still based on the deprecated `go dep` dependency/management at the time of the error.
Indeed goproxy only works with go modules build system.
It would worth to check if we still have the error with the latest version as they have updated their build to use go modules quite recently: https://github.com/gruntwork-io/kubergrunt/pull/95

The changes you are making seems to make sense for go based port which do not use go.mod, it seems less obvious for the on based on go.mod though.
wdyt ?

amake commented:
> The changes you are making seems to make sense for go based port which do not use go.mod, it seems less obvious for the on based on go.mod though.

go.mod does not address the fact that we are relying on upstream resources instead of our own mirrors, which is another point of the MacPorts fetch mechanism.

I am also not sure if it entirely addresses the reproducibility issue: are the fetched dependencies *guaranteed* to be the same at any point in the future? This would not necessarily be the case in the face of things like version range specifications (not sure go.mod supports these) or dependency repos changing their tags up (I guess that should be covered by go.lock, but I'm not positive).

I *know* that the solution proposed in this PR addresses all the issues (at the cost of some maintenance burden).

If you want permission to use go.mod to download dependencies at build time, then that's above my pay grade. Maybe @raimue or @ryandesign can chime in.

amake commented:
I just remembered another reason for the MacPorts distfile fetching policy:

Older systems sometimes *can't* download from upstream because they don't support the minimum TLS version now required by upstream servers. MacPorts mirrors are configured to accommodate older systems like these, so using MacPorts mirroring infrastructure can make the difference between being buildable or not on older systems.

(It may be the case that go only runs on systems new enough to not have this trouble; I'm not sure.)

kencu commented:
> If I understand this doc correctly, we should already use this proxy. Or am I wrong ?
> https://golang.org/cmd/go/#hdr-Module_downloading_and_verification

At first glance, this looks like a great automatic solution to this issue...

dgsb commented:
@amake indeed with the default proxy system of go.mod we would rely on an external proxying system. That's the tradeof right now.

My point is that I'm a bit worried that we are moving from the recommended build system based on module toward an obsolete one. I don't think it's going to be deprecated soon though.

If we do not want to rely on an external proxying system, I really think the long term solution would be to implement our own proxy in the macports infrastructure.

Regarding the build reproduction, that should be covered by the go.sum file. There is no thing like version range in go.mod.

Your solution seems to be quite perfect for go packages which are not based on go modules though.

ryandesign commented:
> I really think the long term solution would be to implement our own proxy in the macports infrastructure.

Please propose it on the macports-dev mailing list so it can be discussed. The discussion doesn't belong in this PR.

amake commented:
> That's the tradeof right now.

There's no need for a tradeoff right now. We have a solution, represented in this PR, that solves all the problems for most golang ports, at the cost of somewhat more difficult maintenance (for which I have provided significant tooling support).

You are proposing a totally new solution, which is fine, but it's hypothetical and in the future.

In the short term I would like to come to a decision one way or another on this PR so I can complete my work on https://trac.macports.org/ticket/61192.

dgsb commented:
@amake the tradeoff is between the infra we are using versus the build system we are using. Sorry but I still don't see the advantage going back to a vendored solution for dependency in the context of building modules based package.
It doesn't seem we do have build error with modules based go package.

dgsb commented:
> > I really think the long term solution would be to implement our own proxy in the macports infrastructure.
> 
> Please propose it on the macports-dev mailing list so it can be discussed. The discussion doesn't belong in this PR.

@ryandesign I will do that when I can find enough time, hopefully today or tomorrow.

amake commented:
> It doesn't seem we do have build error with modules based go package.

I never said there was a build error. I said there is a breach of MacPorts policy as to how distfiles are handled.

> Sorry but I still don't see the advantage going back to a vendored solution for dependency in the context of building modules based package.

I have pointed out the issues several times now; please see [here](https://trac.macports.org/ticket/61192#comment:55) on the off chance that you missed it.

But it seems like you don't "not see the advantage", you just dislike this solution because you perceive the maintenance burden as onerous.

Right now this portfile is not in an acceptable state per MacPorts policy, which exists for the reasons described at the link above. This PR brings the portfile into an acceptable state.

The way forward, then, should be to accept the solution in this PR for the short term, and introduce the better solution as soon as possible.

dgsb commented:
@amake I've checked how the go2port is implemented. I don't think it's correct to compute the dependency from the content of the go.sum files. I've opened an issue in your repository.
